src/blockchain_db/lmdb/db_lmdb.cpp

```diff
@@ -460,49 +471,279 @@ void mdb_txn_safe::allow_new_txns()
   creation_gate.clear();
 }
 
-void lmdb_resized(MDB_env *env)
+mdb_writer_data::~mdb_writer_data()
 {
-  mdb_txn_safe::prevent_new_txns();
+  LOG_PRINT_L3("mdb_writer_data: destructor");
 
-  MGINFO("LMDB map resize detected.");
+  if (m_txn == nullptr)
+    return;
 
-  MDB_envinfo mei;
+  LOG_PRINT_L3("mdb_writer_data: m_txn not nullptr in destructor - calling mdb_txn_abort()");
+  mdb_txn_abort(m_txn);
+}
 
-  mdb_env_info(env, &mei);
-  uint64_t old = mei.me_mapsize;
+boost::thread::id mdb_writer_data::current_thread() const noexcept
+{
+  return m_writer_thread.load(boost::memory_order_relaxed);
+}
 
-  mdb_txn_safe::wait_no_active_txns();
+mdb_databases& mdb_writer_data::databases() noexcept
+{
+  return m_databases;
+}
 
-  int result = mdb_env_set_mapsize(env, 0);
-  if (result)
-    throw0(DB_ERROR(lmdb_error("Failed to set new mapsize: ", result).c_str()));
+mdb_txn_cursors& mdb_writer_data::cursors()
+{
+  return *m_cursors;
+}
 
-  mdb_env_info(env, &mei);
-  uint64_t new_mapsize = mei.me_mapsize;
+MDB_cursor* mdb_writer_data::cursor(source database)
+{
+  if (current_thread() != boost::this_thread::get_id())
+    throw1(DB_ERROR{"transaction owned by other threayyd"});
 
-  MGINFO("LMDB Mapsize increased." << "  Old: " << old / (1024 * 1024) << "MiB" << ", New: " << new_mapsize / (1024 * 1024) << "MiB");
+  auto& cursor = m_cursors->get(database);
 
-  mdb_txn_safe::allow_new_txns();
+  if (cursor != nullptr)
+    return cursor;
+
+  auto res = mdb_cursor_open(m_txn, m_databases.get(database), &cursor);
+
+  if (res == 0)
+    return cursor;
+
+  auto error_message =
+    "Failed to open a cursor for " +
+    mdb_databases::name(database) +
+    ": " +
+    mdb_strerror(res);
+
+  throw0(DB_ERROR(error_message.c_str()));
 }
 
-inline int lmdb_txn_begin(MDB_env *env, MDB_txn *parent, unsigned int flags, MDB_txn **txn)
+MDB_txn* mdb_writer_data::try_acquire() noexcept
 {
-  int res = mdb_txn_begin(env, parent, flags, txn);
-  if (res == MDB_MAP_RESIZED) {
-    lmdb_resized(env);
-    res = mdb_txn_begin(env, parent, flags, txn);
+  if (current_thread() != boost::this_thread::get_id())
+    return nullptr;
+
+  ++m_writer_depth;
+  return m_txn;
+}
+
+MDB_txn* mdb_writer_data::acquire()
+{
+  LOG_PRINT_L3("mdb_writer_data: acquire()");
+
+  if (auto txn = try_acquire())
+    return txn;
+
+  while (mdb_txn_safe::creation_gate.test_and_set());
+  mdb_txn_safe::num_active_txns++;
+  mdb_txn_safe::creation_gate.clear();
+
+  auto result = lmdb_txn_begin(m_env, nullptr, 0, &m_txn);
```

Comment:
Should a temp value `m_txn` value be given to this function? It seems fairly iffy - a second thread is blocked inside of the (eventual) LMDB call but is using the variable that the active writer thread is using. It _probably_ works, assuming LMDB has a full-fence inside somewhere, but ...

Reply:
LMDB write txns are fully serialized, there can be no race here.

Reply:
I know they are serialized, my comment was on re-using the same `&m_txn` in use by the other thread. If LMDB "touches" the memory for this value before the other thread releases the writer lock, then its racy. Maybe LMDB doesn't do that, but this seems to rely on a specific internal implementation.

Reply:
You have just described a race condition. I have just told you there is no race condition; that is a contract of the LMDB API.

Reply:
The documentation doesn't indicate when the out pointer is touched. Sounds like its guaranteed to be after any other thread has completed its work.

Reply:
Of course, because doing *anything* else would constitute a race condition.